File: hstt/scripts/camerad-perf/cpp/run_camerad_expose.cpp
#include <iostream>
#include <vector>
#include <string>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <sstream>
#include <sys/wait.h>
#include <unistd.h>
#include <thread>

std::string executeCommand(const std::string& command) {
    std::string result;
    FILE* pipe = popen(command.c_str(), "r");
    if (!pipe) {
        return "Error: Could not execute command.";
    }
    char buffer[128];
    while (!feof(pipe)) {
        if (fgets(buffer, 128, pipe) != nullptr) {
            result += buffer;
        }
    }
    pclose(pipe);
    return result;
}

int executeCommandWithReturnCode(const std::string& command) {
    int status = system(command.c_str());
    if (WIFEXITED(status)) {
        return WEXITSTATUS(status);
    } else {
        return -1;
    }
}

void executeCommands(const std::vector<std::string>& commands) {
    for (const auto& command : commands) {
        int returnCode = executeCommandWithReturnCode(command);
        if (returnCode == 0) {
            std::cout << "Command '" << command << "' executed successfully" << std::endl;
        } else {
            std::cout << "Command '" << command << "' failed with exit code " << returnCode << "." << std::endl;
        }
    }
}

void executeTimedCommands(const std::vector<std::string>& commands, int n, const std::string& csvFilename = "command_times.csv") {
    std::vector<std::vector<std::string>> results;
    auto startTotal = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < n; ++i) {
        for (const auto& command : commands) {
            auto startTime = std::chrono::high_resolution_clock::now();
            int returnCode = executeCommandWithReturnCode(command);
            auto endTime = std::chrono::high_resolution_clock::now();
            std::chrono::duration<double> duration = endTime - startTime;
            double executionTime = duration.count();

            std::cout << "  Command '" << command << "' executed in " << executionTime << " seconds" << std::endl;
            std::stringstream ss;
            ss << i + 1;
            results.push_back({ss.str(), command, std::to_string(executionTime), std::to_string(returnCode)});
        }
    }

    auto endTotal = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> totalDuration = endTotal - startTotal;
    double totalExecutionTime = totalDuration.count();
    std::cout << "Total time of " << n << " exposures (without prep): " << totalExecutionTime << " seconds." << std::endl;

    std::ofstream csvFile(csvFilename);
    if (csvFile.is_open()) {
        csvFile << "Iteration,Command,Execution Time (seconds),Return Code\n";
        for (const auto& row : results) {
            for (size_t j = 0; j < row.size(); ++j) {
                csvFile << row[j];
                if (j < row.size() - 1) {
                    csvFile << ",";
                }
            }
            csvFile << "\n";
        }
        csvFile.close();
    } else {
        std::cerr << "Unable to open CSV file for writing." << std::endl;
    }
}

int main() {
    std::vector<std::string> prepCommands = {
        "../../camera-interface/bin/socksend -h localhost -p 3031 'open'",
        "../../camera-interface/bin/socksend -h localhost -p 3031 'load'",
        "../../camera-interface/bin/socksend -h localhost -p 3031 'power on'",
        "../../camera-interface/bin/socksend -h localhost -p 3031 'setp Start 1'",
        "../../camera-interface/bin/socksend -h localhost -p 3031 'exptime 0'",
        "../../camera-interface/bin/socksend -h localhost -p 3031 'hsetup'",
        "../../camera-interface/bin/socksend -h localhost -p 3031 'hroi 51 60 51 60'",
        "../../camera-interface/bin/socksend -h localhost -p 3031 'hwindow 1'",
        // "../../camera-interface/bin/socksend -h localhost -p 3031 'zmq 1'",
        "../../camera-interface/bin/socksend -h localhost -p 3031 'autofetch 1'"
    };

    std::vector<std::string> takeExposures = {
        "../../camera-interface/bin/socksend -h localhost -p 3031 'hexpose 5000'"
        // "../../camera-interface/bin/socksend -h localhost -p 3031 'test timer 100 50'"
    };

    executeCommands(prepCommands);
    executeTimedCommands(takeExposures, 1);

    // std::this_thread::sleep_for(std::chrono::seconds( 2 ));
    // executeCommandWithReturnCode("../../camera-interface/bin/socksend -h localhost -p 3031 'abort'");

    return 0;
}

File: hstt/scripts/camerad-perf/py/run_camerad_expose.py
import subprocess
import time
import csv
# import socket

def execute_commands(commands):
    """Execute a list of shell commands."""

    for command in commands:
        try:
            process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
            stdout, stderr = process.communicate()

            if process.returncode == 0:
                print(f"Command '{command}' executed successfully")
            else:
                print(f"Command '{command}' failed with exit code {process.returncode}.")
        except FileNotFoundError:
            print(f"Command '{command}' not found.")
        except Exception as e:
            print(f"An error occurred while executing '{command}': {e}")


def execute_timed_commands(commands, n, csv_filename="command_times.csv"):
    """
    Executes a list of shell commands
    """
    results = []

    start_total = time.time()
    for i in range(n):
        for command in commands:
            start_time = time.time()
            try:
                process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
                stdout, stderr = process.communicate()
                end_time = time.time()
                execution_time = end_time - start_time
                return_code = process.returncode

                print(f"  Command '{command}' executed in {execution_time:.4f} seconds")
                results.append([i + 1, command, execution_time, return_code])
                
            except FileNotFoundError:
                print(f"Command '{command}' not found.")
                results.append([i + 1, command, "Error", "Error"])
            except Exception as e:
                print(f"An error occurred while executing '{command}': {e}")
                results.append([i + 1, command, "Error", "Error"])
                

    end_total = time.time()
    total_execution_time = end_total - start_total
    print(f"Total time of {n} exposures (without prep): {total_execution_time:.4f} seconds.")

    with open(csv_filename, "w", newline="") as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(["Iteration", "Command", "Execution Time (seconds)", "Return Code"])
        writer.writerows(results)

if __name__ == "__main__":

    prep_commands = [
        "../camera-interface/bin/socksend -h localhost -p 3031 'open'",
        "../camera-interface/bin/socksend -h localhost -p 3031 'load'",
        "../camera-interface/bin/socksend -h localhost -p 3031 'power on'",
        "../camera-interface/bin/socksend -h localhost -p 3031 'setp Start 1'",
        "../camera-interface/bin/socksend -h localhost -p 3031 'exptime 0'",
        "../camera-interface/bin/socksend -h localhost -p 3031 'hsetup'",
        "../camera-interface/bin/socksend -h localhost -p 3031 'hroi 100 109 100 109'",
        "../camera-interface/bin/socksend -h localhost -p 3031 'hwindow 1'",
        "../camera-interface/bin/socksend -h localhost -p 3031 'zmq 1'"
        # "../camera-interface/bin/socksend -h localhost -p 3031 'autofetch 1'"
    ]

    take_exposures = [
        "../camera-interface/bin/socksend -h localhost -p 3031 'hexpose 1000'"
    ]

    execute_commands(prep_commands)
    execute_timed_commands(take_exposures, 1)
